lib/Transforms/Instrumentation/ShadowMemSeaDsa.cc

```diff
@@ -877,6 +911,227 @@ void ShadowDsaImpl::updateReadMod(Function &F, NodeSet &readSet,
     // TODO: handle intrinsics (memset,memcpy) and other library functions
   }
 }
+
+bool ShadowDsaImpl::isMemInit(const CallInst &memOp) {
+  assert(memOp.getMetadata(m_metadataTag));
+  auto *fn = memOp.getCalledFunction();
+  assert(fn);
+
+  SmallVector<Constant *, 5> memInitFunctions = {
+      m_memShadowInitFn, m_memGlobalVarInitFn, m_memShadowArgInitFn,
+      m_memShadowUniqArgInitFn, m_memShadowUniqInitFn};
+
+  return llvm::is_contained(memInitFunctions, fn);
+}
+
+CallInst &ShadowDsaImpl::getParentDef(CallInst &memOp) {
+  assert(memOp.getMetadata(m_metadataTag));
+
+  if (isMemInit(memOp))
+    return memOp;
+
+  auto *fn = memOp.getCalledFunction();
+  assert(fn);
+  assert(fn == m_memLoadFn || fn == m_memStoreFn || fn == m_memTrsfrLoadFn);
+
+  assert(memOp.getNumOperands() >= 1);
+  Value *defArg = memOp.getOperand(1);
+  assert(defArg);
+  assert(isa<CallInst>(defArg));
+  auto *def = cast<CallInst>(defArg);
+  auto *defFn = def->getCalledFunction();
+
+  assert(defFn);
+  assert(defFn == m_memStoreFn || defFn == m_argModFn || isMemInit(*def));
+
+  return *def;
+}
+
+const ShadowDsaImpl::MaybeAllocSites &
+ShadowDsaImpl::getAllAllocSites(Value &ptr, AllocSitesCache &cache) {
+  // Performs a simple walk over use-def chains until it finds all allocation
+  // sites or an instruction it cannot look thru (e.g., a load).
+  assert(ptr.getType()->isPointerTy());
+
+  auto *strippedInit = ptr.stripPointerCastsAndBarriers();
+  {
+    auto it = cache.find(strippedInit);
+    if (it != cache.end())
+      return it->second;
+  }
+
+  SmallDenseSet<Value *, 8> allocSites;
+  DenseSet<Value *> visited;
+  SmallVector<Value *, 8> worklist = {strippedInit};
+
+  while (!worklist.empty()) {
+    Value *current = worklist.pop_back_val();
+    assert(current);
+    Value *stripped = current->stripPointerCastsAndBarriers();
+    if (visited.count(stripped) > 0)
+      continue;
+
+    visited.insert(stripped);
+    LOG("shadow_optimizer", llvm::errs() << "Visiting: " << *stripped << "\n");
+
+    // In our memory model, all function calls returning pointers are considered
```

Comment:
what is our memory model? 

why is this a valid thing to do? Will solved form be unsound for a function that simply passes a pointer along?
This is modeled properly in VCGen.

Reply:
We actually have a library routine that determines whether a function returns fresh memory because it is a known library function or one of known allocators.

Reply:
I changed it to check if a function call is a DSA alloc site -- this way we match whatever DSA does already